custom_components/versatile_thermostat/keep_alive.py

```diff
@@ -0,0 +1,53 @@
+"""Building blocks for the heater switch keep-alive feature.
+
+The heater switch keep-alive feature consists of regularly refreshing the state
+of directly controlled switches at a configurable interval (regularly turning the
+switch 'on' or 'off' again even if it is already turned 'on' or 'off'), just like
+the keep_alive setting of Home Assistant's Generic Thermostat integration:
+ https://www.home-assistant.io/integrations/generic_thermostat/
+"""
+
+import logging
+from collections.abc import Awaitable, Callable
+from datetime import timedelta, datetime
+
+from homeassistant.core import HomeAssistant, CALLBACK_TYPE
+from homeassistant.helpers.event import async_track_time_interval
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class IntervalCaller:
+    """Repeatedly call a given async action function at a given regular interval.
+
+    Convenience wrapper around Home Assistant's `async_track_time_interval` function.
+    """
+
+    def __init__(self, hass: HomeAssistant, interval_sec: int) -> None:
+        self._hass = hass
+        self._interval_sec = interval_sec
+        self._remove_handle: CALLBACK_TYPE | None = None
+
+    def cancel(self):
+        """Cancel the regular calls to the action function."""
+        if self._remove_handle:
+            self._remove_handle()
+            self._remove_handle = None
+
+    def set_async_action(self, action: Callable[[], Awaitable[None]]):
+        """Set the async action function to be called at regular intervals."""
+        if not self._interval_sec:
+            return
+        self.cancel()
+
+        async def callback(_time: datetime):
+            try:
+                await action()
+            except Exception as e:  # pylint: disable=broad-exception-caught
+                _LOGGER.error(e)
+                self.cancel()
```

Comment:
The added try/except block avoids ugly job/task stack traces printed in the Home Assistant logs, which would look like a bug in the Versatile Thermostat integration. For example, a raising an unexpected “foo” exception _without_ the try/except block:
```
2024-01-23 23:50:21.545 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved:   File "/home/vscode/.local/bin/hass", line 8, in <module>
    sys.exit(main())
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/__main__.py", line 209, in main
    exit_code = runner.run(runtime_conf)
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/runner.py", line 188, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 640, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 607, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 1914, in _run_once
    handle._run()
  File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/helpers/event.py", line 1454, in _run_async_call_action
    hass.async_run_hass_job(job, time_tracker_utcnow())
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/core.py", line 671, in async_run_hass_job
    hassjob.target(*args)
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/helpers/event.py", line 1532, in interval_listener
    hass.async_run_hass_job(job, now)
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/core.py", line 674, in async_run_hass_job
    return self.async_add_hass_job(hassjob, *args)
  File "/home/vscode/.local/lib/python3.11/site-packages/homeassistant/core.py", line 568, in async_add_hass_job
    task = self.loop.create_task(hassjob.target(*args), name=hassjob.name)
Traceback (most recent call last):
  File "/workspaces/versatile_thermostat/custom_components/versatile_thermostat/keep_alive.py", line 47, in callback
    raise Exception("foo")
Exception: foo
```
And _with_ the try/except block:
```
2024-01-23 23:49:32.513 ERROR (MainThread) [custom_components.versatile_thermostat.keep_alive] foo
```